### ISSUE TITLE
ABRP compatibility

### ISSUE THREAD

**Issue by teodorch85** [meatpiHQ/wican-fw#179]
I would like to ask if someone has used WiCan with ABRP
https://www.iternio.com/abrp-obd

**Comment by jay-oswald** [comment]
No i would like to. But you can't currently have station mode and bluetooth active at the same time. I have an issue about this, that is planned to fix this, then should he able to activate bluetooth and use with abrp.

**Comment by meatpiHQ** [comment]
So if I understand correctly, you need BLE to be ON during when WiCAN is not connected to wifi. And once BLE disconnected, supposedly when you arrive home you want it to connect to you home wifi. 

The only issue here is that you have to manually disconnect from BLE.

I had issues with BLE and WiFi coexisting on the same radio, it significantly throttled the connection but that was a while ago I have to retest with the new idf to see if this was fixed and how well it works.

**Comment by jay-oswald** [comment]
@meatpiHQ I don't think manually disconnecting from BLE is a major issue. Though Ideally do something smart like if ignition has been off for 2+ mins turn off the BLE.

I'd be fine with an option to only have BLE enabled when station can't connect, and then keep station disabled while BLE is connected, especailly if BLE automatically drops the connection when the car is turned off for a few mins

**Comment by DeepflashX** [comment]
Hello all togetheter,
how is the status regarding ABRP?
I would be interested in the MeatPi product if ABRP would work and have Wifi running aswell.

**Comment by meatpiHQ** [comment]
@DeepflashX It's not yet implemented. I will update this issue once it's ready.

> I would be interested in the MeatPi product if ABRP would work and have Wifi running aswell.

Do you connect ABRP to BLE and have WiFi connected at the same time? 

This is possible but it affect the throughput, I have this enabled at some point but it did not work so well.

**Comment by DeepflashX** [comment]
@meatpiHQ ABRP would use BLE to connect to the Dongle. I would set up a Hotspot on my Android Phone and let the Dongle connect to it and upload to MQTT. So yes, Wifi and BLE would be needed same time but of course only a limted number of PIDs.
In the end I need a maximum of 20 PIDs to upload to MQTT.

**Comment by cr0i** [comment]
I am also interested in using ABRP with WiCAN.
I have tried disabling WLAN Station Mode so that I can enable Bluetooth (I hope BLE?) in WiCAN.
With my iPhone, the WiCAN is not found as a Bluetooth device.
My Android has found it, but when I connect it, it immediately disconnects.
No device is found in ABRP.

Has anyone managed to get it to work with ABRP at all?

I only want to use ABRP on the go, at home I only need a WLAN connection.
In my use case, BLE is therefore only needed if the WLAN connection is interrupted.

**Comment by cr0i** [comment]
Hello @meatpiHQ ,

I had to change the Protocoll from AutoPID (I used with WLAN) to elm327 (when trying BLE), so connection with ABRP was successfully.
This could be important for this solution.

**Comment by meatpiHQ** [comment]
@cr0i Thanks. I want to add a "Drive mode" and "Home Mode", so when the car is running it will activate BLE and when it's parked at home it will connect to home wifi.

**Comment by DeepflashX** [comment]
Both at the same time is not possible?
My plan was to set up a Hotspot on my mobile and have both ABRP running and wifi to upload to Mqtt..

**Comment by mdallaire** [comment]
I am looking at the Wican Pro for my newer car but I still use a OVMS with LTE on my old Nissan Leaf and the way to link it to ABRP is using https://github.com/iternio/ovms-link to push the vehicle metrics to the ABRP api. 

Perhaps this is something that can be adapted to the Wican code? This way the metrics are pushed through the network (either cellphone hotspot or the future LTE dongle) and there is no need to enable the Bluetooth stack.

**Comment by meatpiHQ** [comment]
@mdallaire thank you for sharing this

**Comment by juggie** [comment]
@meatpiHQ Does ABRP work with BLE with WICAN-PRO?  Not sure what the current status is, I've not been able to find the information.  I don't need BLE + WIFI to work in parallel (though perhaps you could get both working if you used BLE + WIFI5G only, ie not 2.4G which is where it collides with Bluetooth).

I was looking to get one dongle for both ABRP, but also playing with some home assistant and android auto stuff but i don't need it all to work in parallel.  Sorry if this isn't a good place to ask this, it was the best location I could find.

**Comment by meatpiHQ** [comment]
@juggie Yes WICAN-PRO works with ABRP

**Comment by gerok1984** [comment]
![Image](https://github.com/user-attachments/assets/ac8988c8-e434-4abd-b317-a28b18a3a1bf)

What settings do you use to make WiCan work with ABRP?
I've enabled BLE and selected ELM327, and it says Not Compatible.

![Image](https://github.com/user-attachments/assets/c7fe4791-ade0-403e-9252-ecdc83424e58)

**Comment by meatpiHQ** [comment]
@gerok1984 for now only the PRO version is compatible with ABRP.

**Comment by gerok1984** [comment]
> [@gerok1984](https://github.com/gerok1984) for now only the PRO version is compatible with ABRP.

Ohh. I was misinformed then. I canceled my order for the Pro version last month thinking the regular one would work, and now there are no units available of PRO until October in CrowdSupply. 😅

**Comment by meatpiHQ** [comment]
@gerok1984 More stock should be available in few weeks. The PRO version is much easier to develop on and will be getting more advanced features, I would recommend it over the original WiCAN.

**Comment by Manouchehri** [comment]
It'd be great if the WiCAN Pro could push data to the ABRP API, without needing to use BLE at all. 🙂

**Comment by juggie** [comment]
> It'd be great if the WiCAN Pro could push data to the ABRP API, without needing to use BLE at all. 🙂

Interesting thought but the wican has no way to do that.  The only option would be to put the wican in AP mode, and you could connect your android phone to it, but then your phone could not use Android Auto/Carplay and would not have internet so I suspect that would not be desirable.   

I believe you may also be referencing the ABRP Token process where you can push data to ABRP w/ your token to create a link to your car.  The only way I could see this working is if you had the LTE modem for your WiCAN-PRO.

**Comment by Manouchehri** [comment]
> The only way I could see this working is if you had the LTE modem for your WiCAN-PRO.

Why not use your phone as a hotspot?

> but then your phone could not use Android Auto/Carplay

You can just use wired CarPlay / Android Auto and then your phone can be used as a hotspot for the WiCAN Pro.

> I believe you may also be referencing the ABRP Token process where you can push data to ABRP w/ your token to create a link to your car.

Correct.

**Comment by juggie** [comment]
> > The only way I could see this working is if you had the LTE modem for your WiCAN-PRO.
> 
> Why not use your phone as a hotspot?
> 
> > but then your phone could not use Android Auto/Carplay
> 
> You can just use wired CarPlay / Android Auto and then your phone can be used as a hotspot for the WiCAN Pro.
> 
> > I believe you may also be referencing the ABRP Token process where you can push data to ABRP w/ your token to create a link to your car.
> 
> Correct.

Good point, I didn't think of that direction.  There was already some talk of a http/post functionality, this could possibly solve your use case, not sure if that's still a priority.  @meatpiHQ would be able to comment.

**Comment by meatpiHQ** [comment]
I'm working on adding http/https , it's not specific to ABRP. Does ABRP expect a specific data format ?

**Comment by Manouchehri** [comment]
> I'm working on adding http/https , it's not specific to ABRP. Does ABRP expect a specific data format ?

If you can POST to any HTTP(S) endpoint with all the CAN data, that's enough for me, as I can write my own Cloudflare Worker to push the data wherever I want. :)

**Comment by meatpiHQ** [comment]
@Manouchehri the way OBD protocol works, is that you have to request the data from the vehicle ECU. So basically you send a CAN frame listen to the response and interpret the results.

**Comment by stevelea** [comment]
> I'm working on adding http/https , it's not specific to ABRP. Does ABRP expect a specific data format ?

Just as described here https://documenter.getpostman.com/view/7396339/SWTK5a8w

**Comment by meatpiHQ** [comment]
@Manouchehri @stevelea ABRP api is now supported please test and let me know if it works for you. I did not test with ABRP, I don't have a token yet. I've only tested on my own python server.

This new firmware will allow you to use BLE+Station mode and also support HTTPs post, you can upload certificates in the system tab if needed.

[wican-fw_obd_pro_v430p_beta5.zip](https://github.com/user-attachments/files/22236943/wican-fw_obd_pro_v430p_beta5.zip)

**Comment by cr0i** [comment]
> [@Manouchehri](https://github.com/Manouchehri) [@stevelea](https://github.com/stevelea) ABRP api is now supported please test and let me know if it works for you. I did not test with ABRP, I don't have a token yet. I've only tested on my own python server.
> 
> This new firmware will allow you to use BLE+Station mode and also support HTTPs post, you can upload certificates in the system tab if needed.
> 
> [wican-fw_obd_pro_v430p_beta5.zip](https://github.com/user-attachments/files/22236943/wican-fw_obd_pro_v430p_beta5.zip)

Hello @meatpiHQ ,
will there be an Update for WiCAN (without Pro) too, so we can use BLE one the Road, when WiFi is not connected?
This would not only be nice for ABPR, so I would also be able to use Carscanner with CarPlay/Android Auto.

**Comment by stevelea** [comment]
> [@Manouchehri](https://github.com/Manouchehri) [@stevelea](https://github.com/stevelea) ABRP api is now supported please test and let me know if it works for you. I did not test with ABRP, I don't have a token yet. I've only tested on my own python server.
> 
> This new firmware will allow you to use BLE+Station mode and also support HTTPs post, you can upload certificates in the system tab if needed.
> 
> [wican-fw_obd_pro_v430p_beta5.zip](https://github.com/user-attachments/files/22236943/wican-fw_obd_pro_v430p_beta5.zip)

Loaded up Beta5, i cant update the network configuration as it is greyed out, i can add an additional ssid but it doesnt appear to switch to that.  Also i currenty use homeassistant to fill the data for ABRP and my car profile does not appear to support BT Connection in ABRP. When i switch on BT i get a message saying AP Will disconnect, which it doesnt BTW.

**Comment by meatpiHQ** [comment]
@stevelea can you please send me a screenshot of the greyed out network configuration?

Can you confirm you have the correct security configuration wpa2 or wpa3 in the other ssid?

**Comment by stevelea** [comment]
> [@stevelea](https://github.com/stevelea) can you please send me a screenshot of the greyed out network configuration?
> 
> Can you confirm you have the correct security configuration wpa2 or wpa3 in the other ssid?

User issue :-) .. ignore the comment :-)

**Comment by stevelea** [comment]
> [@Manouchehri](https://github.com/Manouchehri) [@stevelea](https://github.com/stevelea) ABRP api is now supported please test and let me know if it works for you. I did not test with ABRP, I don't have a token yet. I've only tested on my own python server.
> 
> This new firmware will allow you to use BLE+Station mode and also support HTTPs post, you can upload certificates in the system tab if needed.
> 
> [wican-fw_obd_pro_v430p_beta5.zip](https://github.com/user-attachments/files/22236943/wican-fw_obd_pro_v430p_beta5.zip)

Just testing abrp , so in the url field for abrp_api i have https://api.iternio.com/1/tlm/send? and then in the api field i have apikey=c8b6c1b3-72be-4706-b5ea-xxxxxxxxxxxc&token=88b3baad-0b6a-4308-b6e4-xxxxxxxxxxx  , is that the format you expect ? here is my rest command i use in HomeAssistant which work btw 

 abrp:
    url: >
      https://api.iternio.com/1/tlm/send?token={{ token | default('MISSING_TOKEN') }}&tlm={
        "utc":{{ utc | default(0) }},
        "soc":{{ soc | default(0) }},
        "soh":{{ soh | default(100) }},
        "power":{{ power | default(0) }},
        "speed":"{{ speed | default('') }}",
        "lat":{{ lat | default(0) }},
        "lon":{{ lon | default(0) }},
        "is_charging":{{ is_charging | default(0) }},
        "is_dcfc":{{ is_dcfc | default(0) }},
        "is_parked":{{ is_parked | default(0) }},
        "elevation":"{{ elevation | default('') }}",
        "ext_temp":"{{ ext_temp | default('') }}",
        "odometer":{{ odometer | default(0) }},
        "est_battery_range":{{ est_battery_range | default(0) }}
      }
    method: post
    headers:
      Authorization: "APIKEY {{ api_key }}"

I think its working, let me just wait a while as i have turned off my HA automation for the moment

**Comment by meatpiHQ** [comment]
@stevelea I think I'll have to add another field for API key

**Comment by stevelea** [comment]
> [@stevelea](https://github.com/stevelea) I think I'll have to add another field for API key

It does appear to be working

**Comment by stevelea** [comment]
> > [@stevelea](https://github.com/stevelea) I think I'll have to add another field for API key
> 
> It does appear to be working 

Yes confirmed working using the string I had in the above comment

**Comment by stevelea** [comment]
Using 430p_beta5 the dialog to specficy the address of the mqtt server seems to have disappeared , or i am going blind :-)

**Comment by r1st0** [comment]
> > [@Manouchehri](https://github.com/Manouchehri) [@stevelea](https://github.com/stevelea) ABRP api is now supported please test and let me know if it works for you. I did not test with ABRP, I don't have a token yet. I've only tested on my own python server.
> > This new firmware will allow you to use BLE+Station mode and also support HTTPs post, you can upload certificates in the system tab if needed.
> > [wican-fw_obd_pro_v430p_beta5.zip](https://github.com/user-attachments/files/22236943/wican-fw_obd_pro_v430p_beta5.zip)
> 
> Hello [@meatpiHQ](https://github.com/meatpiHQ) , will there be an Update for WiCAN (without Pro) too, so we can use BLE one the Road, when WiFi is not connected? This would not only be nice for ABPR, so I would also be able to use Carscanner with CarPlay/Android Auto.

@meatpiHQ will ABRP support also come to non-pro version (WICAN OBD)? or more generally, will there come further releases for the wican obd in the future?

**Comment by meatpiHQ** [comment]
@r1st0 Yes there will be future firmware releases for WiCAN-OBD. I plan to try and add support for ABRP, I don't know how well it's going to work.

**Comment by alistair23** [comment]
It seems like the latest release has ABRP support in the UI, but it doesn't seem to work as the data doesn't update in ABRP

<img width="1629" height="952" alt="Image" src="https://github.com/user-attachments/assets/7693bfac-9768-481b-adfa-d5ff06170cf2" />

**Comment by stevelea** [comment]
> It seems like the latest release has ABRP support in the UI, but it doesn't seem to work as the data doesn't update in ABRP
> 
> <img width="1629" height="952" alt="Image" src="https://github.com/user-attachments/assets/7693bfac-9768-481b-adfa-d5ff06170cf2" />

Do you have both the apikey and token in  key field ? Example  apikey=c8b6c1b3-72be-4706-b5ea-xxxxxxxxxxxc&token=88b3baad-0b6a-4308-b6e4-xxxxxxxxxxx

**Comment by gerok1984** [comment]
> > It seems like the latest release has ABRP support in the UI, but it doesn't seem to work as the data doesn't update in ABRP
> > <img alt="Image" width="1629" height="952" src="https://private-user-images.githubusercontent.com/171674/506529093-7693bfac-9768-481b-adfa-d5ff06170cf2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE2NTUwODksIm5iZiI6MTc2MTY1NDc4OSwicGF0aCI6Ii8xNzE2NzQvNTA2NTI5MDkzLTc2OTNiZmFjLTk3NjgtNDgxYi1hZGZhLWQ1ZmYwNjE3MGNmMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyOFQxMjMzMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wODljNzU1NTA5YTA3M2U4MTU0MzQ0NWYxZjc2MTg1ZjBjYmZhZmQzNzFjZDNjZDQ2NzNiNjA5YzU1Y2I3MTM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DMIR1w3Lr1JWh6SwOz2x-dpsKShCcrlNrmkraRSI2vc">
> 
> Do you have both the apikey and token in key field ? Example apikey=c8b6c1b3-72be-4706-b5ea-xxxxxxxxxxxc&token=88b3baad-0b6a-4308-b6e4-xxxxxxxxxxx

![Image](https://github.com/user-attachments/assets/1e3280e2-9443-497b-8dcb-15dec331ef1b)

Is this the correct format?

**Comment by stevelea** [comment]
> > > It seems like the latest release has ABRP support in the UI, but it doesn't seem to work as the data doesn't update in ABRP
> > > <img alt="Image" width="1629" height="952" src="https://private-user-images.githubusercontent.com/171674/506529093-7693bfac-9768-481b-adfa-d5ff06170cf2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE2NTUwODksIm5iZiI6MTc2MTY1NDc4OSwicGF0aCI6Ii8xNzE2NzQvNTA2NTI5MDkzLTc2OTNiZmFjLTk3NjgtNDgxYi1hZGZhLWQ1ZmYwNjE3MGNmMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyOFQxMjMzMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wODljNzU1NTA5YTA3M2U4MTU0MzQ0NWYxZjc2MTg1ZjBjYmZhZmQzNzFjZDNjZDQ2NzNiNjA5YzU1Y2I3MTM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DMIR1w3Lr1JWh6SwOz2x-dpsKShCcrlNrmkraRSI2vc">
> > 
> > Do you have both the apikey and token in key field ? Example apikey=c8b6c1b3-72be-4706-b5ea-xxxxxxxxxxxc&token=88b3baad-0b6a-4308-b6e4-xxxxxxxxxxx
> 
> ![Image](https://github.com/user-attachments/assets/1e3280e2-9443-497b-8dcb-15dec331ef1b)
> 
> Is this the correct format? 

Yes looks ok

**Comment by gerok1984** [comment]
It doesn't work for me.
It doesn't send SOC data or any other info.
It does work from Home Assistant.

I also don't know why it won't let me activate MQTT HA Discovery. It os in
grey.

**Comment by meatpiHQ** [comment]
@gerok1984 MQTT discovery cannot be enabled I should remove it from the UI. You should use the HA integration

**Comment by spuky** [comment]
@gerok1984 Docs state 

"Grouping Enabled | Automate top section → Grouping = Enable"

it is disabled in you screenshots..

**Comment by camueller** [comment]
With 4.45 ABRP works out-of-the-box without token etc.

![Image](https://github.com/user-attachments/assets/89bfd38b-a518-4986-b7f9-4cf2465e9389)

![Image](https://github.com/user-attachments/assets/f338fff9-b8fa-45c4-8659-729db7e85de5)

**Comment by evcc-passra** [comment]
> [@cr0i](https://github.com/cr0i) Thanks. I want to add a "Drive mode" and "Home Mode", so when the car is running it will activate BLE and when it's parked at home it will connect to home wifi.

is this feature now added at the WiCan-pro?

**Comment by vanrace** [comment]
> > [@cr0i](https://github.com/cr0i) Thanks. I want to add a "Drive mode" and "Home Mode", so when the car is running it will activate BLE and when it's parked at home it will connect to home wifi.
> 
> is this feature now added at the WiCan-pro?

Interested in this as well!
